src/primitives/iter.rs

```diff
@@ -0,0 +1,501 @@
+    ///
+    /// This function is infallible, but if you feed it a non-ASCII `hrp` it probably
+    /// will cause your checksum engine to produce useless results.
+    pub fn with_checksummed_hrp(mut self, hrp: &Hrp) -> Self {
+        self.checksum_engine.input_hrp(hrp);
+        self
+    }
+
+    /// Helper function to input an extra field element into the underling
+    /// checksum engine of the iterator.
+    pub fn with_checksummed_fe(mut self, fe: Fe32) -> Self {
+        self.checksum_engine.input_fe(fe);
+        self
+    }
+}
+
+impl<I, Ck> Iterator for ChecksumIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    type Item = Fe32;
+
+    fn next(&mut self) -> Option<Fe32> {
+        match self.iter.next() {
+            Some(fe) => {
+                self.checksum_engine.input_fe(fe);
+                Some(fe)
+            }
+            None =>
+                if self.checksum_remaining == 0 {
+                    None
+                } else {
+                    if self.checksum_remaining == Ck::CHECKSUM_LENGTH {
+                        self.checksum_engine.input_target_residue();
+                    }
+                    self.checksum_remaining -= 1;
+                    Some(Fe32(self.checksum_engine.residue().unpack(self.checksum_remaining)))
+                },
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.iter.size_hint();
+        (min + self.checksum_remaining, max.map(|max| max + self.checksum_remaining))
+    }
+}
+
+/// Iterator adaptor which takes a stream of field elements, converts it to characters prefixed by
+/// an HRP. If `fe_iter` is a checksummed iter, it is expected that the `hrp` strings are identical.
+pub struct HrpCharIter<'hrp, I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    hrp_iter: hrp::LowercaseCharIter<'hrp>,
+    fe_iter: I,
+    hrp_done: bool,
+}
+
+impl<'hrp, I> Iterator for HrpCharIter<'hrp, I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    type Item = char;
+
+    fn next(&mut self) -> Option<char> {
+        if !self.hrp_done {
+            match self.hrp_iter.next() {
+                Some(c) => return Some(c),
+                None => {
+                    self.hrp_done = true;
+                    return Some('1');
+                }
+            }
+        }
+        self.fe_iter.next().map(Fe32::to_char)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.fe_iter.size_hint();
+        let add = if !self.hrp_done {
+            self.hrp_iter.len() + 1 // hrp | SEP
+        } else {
+            0
+        };
+
+        (min + add, max.map(|max| max + add))
+    }
+}
+
+/// Returns the number of Fe32's encoded by `n` bytes.
+fn bytes_len_to_fes_len(n: usize) -> usize {
+    let bits = n * 8;
+    if bits % 5 != 0 {
+        bits / 5 + 1 // +1 because we pad the fes.
+    } else {
+        bits / 5
+    }
```

Comment:
In 0c6613106fa624db037598dab5e9a19634acd2d0:

It would be clearer and more idiomatic to just write `(bits + 4) / 5` if you mean "bits / 5, rounded up". TBH I think this function could be replaced by inline code and doesn't need any unit tests.

Reply:
Done, same for the other direction (fes to bytes)